inspectit-oce-core/src/test/java/rocks/inspectit/oce/core/instrumentation/dataprovider/DataProviderGeneratorIntTest.java

```diff
@@ -0,0 +1,235 @@
+    static void initDummyClassloader() throws Exception {
+        dummyLoader = new DummyClassLoader(DataProviderGeneratorIntTest.class.getClassLoader(), MyDummyClass.class);
+        dummyClass = Class.forName(MyDummyClass.class.getName(), false, dummyLoader);
+    }
+
+    IGenericDataProvider getInstance(InjectedClass<? extends IGenericDataProvider> clazz) {
+        try {
+            return (IGenericDataProvider) clazz.getInjectedClassObject().get().getField("INSTANCE").get(null);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    @DirtiesContext
+    void testSyntaxErrorHandling() {
+        ResolvedGenericDataProviderConfig config = ResolvedGenericDataProviderConfig.builder()
+                .name("my-provider")
+                .valueBody("i am not java code")
+                .build();
+
+        InjectedClass<?> provider = generator.getOrGenerateDataProvider(config, dummyClass);
+        assertThat(provider.getInjectedClassObject().get()).isSameAs(GenericDataProviderTemplate.class);
+    }
+
+    @Test
+    @DirtiesContext
+    void testProviderCaching() {
+        ResolvedGenericDataProviderConfig config = ResolvedGenericDataProviderConfig.builder()
+                .name("my-provider")
+                .valueBody("return \"dummyresult\";")
+                .build();
+
+        InjectedClass<? extends IGenericDataProvider> providerA = generator.getOrGenerateDataProvider(config, dummyClass);
+        assertThat(getInstance(providerA).execute(null, null, null, null, null)).isEqualTo("dummyresult");
+
+        InjectedClass<? extends IGenericDataProvider> providerB = generator.getOrGenerateDataProvider(config, dummyClass);
+        assertThat(providerB).isSameAs(providerA);
+    }
+
+
+    @Test
+    @DirtiesContext
+    void testProviderReuse() throws Exception {
+        ResolvedGenericDataProviderConfig configA = ResolvedGenericDataProviderConfig.builder()
+                .name("my-provider")
+                .valueBody("return \"providerA\";")
+                .build();
+
+        InjectedClass<? extends IGenericDataProvider> providerA = generator.getOrGenerateDataProvider(configA, dummyClass);
+        assertThat(getInstance(providerA).execute(null, null, null, null, null)).isEqualTo("providerA");
+
+        Class<?> providerAClass = providerA.getInjectedClassObject().get();
+        WeakReference<Object> providerAWeak = new WeakReference<>(providerA);
+        providerA = null;
+
+        GcUtils.waitUntilCleared(providerAWeak);
+
+        ResolvedGenericDataProviderConfig configB = ResolvedGenericDataProviderConfig.builder()
+                .name("my-provider")
+                .valueBody("return \"providerB\";")
+                .build();
+
+        InjectedClass<? extends IGenericDataProvider> providerB = generator.getOrGenerateDataProvider(configB, dummyClass);
+
+        assertThat(providerB.getInjectedClassObject().get()).isSameAs(providerAClass);
```

Comment:
execute to confirm result is `providerB`?

Reply:
it will actually return "providerA" because redefineClasses is mocked in the integration test.